BottomAppBar: Fix doubled layers of color and shadow

chrisbobbe opened the pull request:
Fixes #123283.
Fixes #122667.
Fixes #123064.

(EDITED for the current revision, following https://github.com/flutter/flutter/issues/123283#issuecomment-1481637735.)

The `BottomAppBar` implementation uses both a `Material` and a `PhysicalShape` because it needs them both in order to support the "notch" for a FAB in Material 2.

In the current version of the Material 3 support, it ends up putting a color and shadow on both of these, so that there are two color layers and two shadows. This gives wrong results when the color is translucent: #122667, #123064.

Also in Material 3, it doesn't work right if you ask for a notch, because the `Material` paints through the notch: #123283. (The Material 3 [spec](https://m3.material.io/components/bottom-app-bar/specs) doesn't allow for a notch in the bottom app bar, but we plan to support one anyway: https://github.com/flutter/flutter/issues/123283#issuecomment-1481637735.)

So, fix these bugs by making the `Material` transparent (and without overlay or shadow, using `MaterialType.transparency`), like we do in Material 2.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

chrisbobbe committed:
BottomAppBar: Explicitly unsupport notched `shape` with Material 3

Fixes #123283.
Fixes #122667.
Fixes #123064.

The Material 3 spec removes the "notched" shape for bottom app bars
that was present Material 2:
  https://m3.material.io/components/bottom-app-bar/specs

So, remove the `PhysicalShape` widget that was providing that
notched shape, and throw if callers in Material 3 ask for a notch.
Callers expecting a notch with Material 3 will have been
disappointed since da7b8327e, when the `Material` widget (which
doesn't support a notch) started painting its background color
through any notch provided by the `PhysicalShape` widget. This is
issue #123283.

As a bonus, this fixes some bugs observed in Material 3 that were
invited by BottomAppBar's two-layer implementation (the
`PhysicalShape` and the `Material`):

- #122667, "BottomAppBar: Material 3 makes the color too opaque"
- #123064, "BottomAppBar: Elevation effect (shadow) doubled in
  Material 3"

chrisbobbe committed:
BottomAppBar: Refactor nested-child pattern to use fewer variable names

chrisbobbe committed:
BottomAppBar: Refactor to consolidate `isMaterial3` conditionals

It's now a bit easier to read complete pictures of the Material 3
behavior and of the Material 2 behavior.

Soon we'll remove the PhysicalShape widget from the Material 3 case
entirely -- its job isn't needed in Material 3 because the bottom
app bar's "notched" shape isn't part of the spec anymore:
  https://m3.material.io/components/bottom-app-bar/specs

chrisbobbe committed:
BottomAppBar: Remove some redundant arg values that match defaults

chrisbobbe committed:
BottomAppBar: Refactor to tighten up code

chrisbobbe committed:
BottomAppBar: Remove a no-op `surfaceTintColor` in Material 2 case

`Material.surfaceTintColor` is ignored in Material 2. From the doc:

https://github.com/flutter/flutter/blob/2ad6cd72c0/packages/flutter/lib/src/material/material.dart#L283
> If [ThemeData.useMaterial3] is false, then this property is not used.

It was added for Material 3 support in 6a66aa282.

chrisbobbe committed:
BottomAppBar: Stop redundantly giving `elevation` / shadowColor to Material

We already pass these to the `PhysicalShape`; the transparent
Material doesn't need them too.

As noted in
  https://github.com/flutter/flutter/issues/123064#issuecomment-1477050713
, I don't think this redundancy in the Material 2 case has actually
caused a bug where an elevation effect gets doubled, because:
- A `Material` with `type: MaterialType.transparency` doesn't
  display a shadow:
    https://github.com/flutter/flutter/blob/f4c3facfd/packages/flutter/lib/src/material/material.dart#L522-L529
- `BottomAppBar` tries to apply an elevation overlay effect using
  `color` (with `ElevationOverlay.applyOverlay`), but `color` gets
  ignored by `Material` if `type` is `MaterialType.transparency`:
    https://github.com/flutter/flutter/blob/f4c3facfd/packages/flutter/lib/src/material/material.dart#L522-L529

Still, that's not trivial to verify, and it's more confusing to
keep passing these arguments.

(Note that this addresses the Material 2 case. In a recent commit,
we removed the `PhysicalShape` in the Material 3 case, fixing some
live bugs where some effects had in fact been duplicated by the
presence of the `PhysicalShape` together with the `Material`.)

chrisbobbe committed:
BottomAppBar: Inline `effectiveColor` in its two use sites

chrisbobbe committed:
BottomAppBar: Remove unused `shape` default in Material 3 case

Done by removing the line in bottom_app_bar_template.dart then:

  dart dev/tools/gen_defaults/bin/gen_defaults.dart

and undoing some unrelated-looking changes that made in these files:

  packages/flutter/lib/src/material/bottom_sheet.dart
  packages/flutter/lib/src/material/search_anchor.dart

chrisbobbe commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -191,34 +208,38 @@ class _BottomAppBarState extends State<BottomAppBar> {
+        child: SafeArea(child: child),
+      );
+    } else {
+      child = PhysicalShape(
+        clipper: clipper,
+        elevation: elevation,
+        shadowColor: shadowColor,
+        color: ElevationOverlay.applyOverlay(context, color, elevation),
+        clipBehavior: widget.clipBehavior,
+        child: Material(
+          key: materialKey,
+          type: MaterialType.transparency,
+          child: SafeArea(child: child),
```
This is exactly like how this code was before Material 3 support was first added in #106525, except for the customizations added for padding in #115175 and for shadow color in #121406.

chrisbobbe committed:
BottomAppBar tests: Fill in missing issue number #123283

gnprice commented on packages/flutter/test/material/bottom_app_bar_test.dart:
```diff
@@ -11,7 +11,88 @@ import 'package:flutter/material.dart';
+import '../rendering/mock_canvas.dart';
+
 void main() {
+  // Regression test for https://github.com/flutter/flutter/issues/123283
+  testWidgets('throws if `shape` requested in Material 3', (WidgetTester tester) async {
```
nit:
```suggestion
  testWidgets('throws if `shape` requested in Material 3', (WidgetTester tester) async {
    // Regression test for https://github.com/flutter/flutter/issues/123283
```

and similarly below

gnprice commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -174,6 +189,8 @@ class _BottomAppBarState extends State<BottomAppBar> {
   Widget build(BuildContext context) {
     final ThemeData theme = Theme.of(context);
     final bool isMaterial3 = theme.useMaterial3;
+    assert(!isMaterial3 || widget.shape == null); // `shape` not accepted in Material 3
```
The information in the comment should go in an assertion message instead, so the developer sees it when they hit the assertion. For example:

```suggestion
    assert(!isMaterial3 || widget.shape == null,
      'The [BottomAppBar.shape] field is not permitted in Material 3.');
```

(Could also throw a `FlutterError` for a fancier error message, but this might be enough.)

chrisbobbe replied:
If throwing a `FlutterError`, would we want to do that in production too? AIUI [`assert`s only run in debug mode](https://docs.flutter.dev/testing/debugging#debug-mode-assertions).

gnprice replied:
No, the check and the throw would both go inside an `assert`. For example, here's [State.context]:
```dart
  BuildContext get context {
    assert(() {
      if (_element == null) {
        throw FlutterError(
          'This widget has been unmounted, so the State no longer has a context (and should be considered defunct). \n'
          'Consider canceling any active work during "dispose" or using the "mounted" getter to determine if the State is still active.',
        );
      }
      return true;
    }());
    return _element!;
  }
```

gnprice commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -114,17 +117,29 @@ class BottomAppBar extends StatefulWidget {
   /// {@macro flutter.material.Material.clipBehavior}
   ///
   /// Defaults to [Clip.none], and must not be null.
+  ///
+  /// Ignored if the ambient [ThemeData.useMaterial3] is true; the Material 3
+  /// spec defines the shape without a notch:
+  ///   https://m3.material.io/components/bottom-app-bar/specs#9d91e4a4-a6d9-4ad5-afad-430267ebb4e9
   final Clip clipBehavior;
```
Can this be simplified like so?

```suggestion
  ///
  /// Has no effect if there is no notch; see [shape].
  final Clip clipBehavior;
```

Similarly for [notchMargin].

(And if so, then that seems like a good docs improvement even before this PR's changes to the behavior of the code.)

gnprice commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -114,17 +117,29 @@ class BottomAppBar extends StatefulWidget {
   /// [ThemeData.bottomAppBarTheme] is used. If that's null then the shape will
   /// be rectangular with no notch.
+  ///
+  /// Ignored if the ambient [ThemeData.useMaterial3] is true; the Material 3
+  /// spec defines the shape without a notch:
+  ///   https://m3.material.io/components/bottom-app-bar/specs#9d91e4a4-a6d9-4ad5-afad-430267ebb4e9
   final NotchedShape? shape;
```
This isn't just ignored in that case, but rather not permitted, right? So for example:

```suggestion
  /// If the ambient [ThemeData.useMaterial3] is true,
  /// then there is never a notch, and this property must be null.
  /// (The Material 3 spec does not provide for a notch:
  /// <https://m3.material.io/components/bottom-app-bar/specs#9d91e4a4-a6d9-4ad5-afad-430267ebb4e9>.)
  final NotchedShape? shape;
```

gnprice reviewed (commented):
Thanks for the fix! Generally this looks good.

chrisbobbe committed:
BottomAppBar tests: Reorder some "regression test" comments

chrisbobbe committed:
BottomAppBar: Add developer-facing error message in no-notch `assert`

chrisbobbe committed:
BottomAppBar: Adjust docs from gnprice's review

chrisbobbe committed:
BottomAppBar: Cut legacy "must not be null" comments from pre-null-safety

chrisbobbe commented:
Thanks for the review, @gnprice! Changes pushed; PTAL. 🙂

gnprice commented on packages/flutter/test/material/bottom_app_bar_test.dart:
```diff
@@ -29,8 +29,8 @@ void main() {
 
-  // Regression test for https://github.com/flutter/flutter/issues/123064
   testWidgets('shadow effect is not doubled', (WidgetTester tester) async {
+  // Regression test for https://github.com/flutter/flutter/issues/123064
```
nit: indentation (here and below)
```suggestion
  testWidgets('shadow effect is not doubled', (WidgetTester tester) async {
    // Regression test for https://github.com/flutter/flutter/issues/123064
```

gnprice reviewed (approved):
LGTM! One nit.

chrisbobbe committed:
BottomAppBar tests: Fix indentation nits

chrisbobbe commented:
Thanks for the review! Fixed pushed.

chrisbobbe committed:
BottomAppBar: Re-enable "notched" shape in Material 3, but make it work

Following HansMuller's comment at
  https://github.com/flutter/flutter/issues/123283#issuecomment-1481637735

> Yes, I mean that the notch should still work with
> `useMaterial3: true`.

chrisbobbe commented:
I'm reworking this following @HansMuller's comment at https://github.com/flutter/flutter/issues/123283#issuecomment-1481637735:

> [T]he notch should still work with `useMaterial3: true`.

justinmc commented:
Sounds good, I'm happy to give a secondary review when this is back out of draft mode.

chrisbobbe committed:
BottomAppBar tests: Make some element-tree expectations more explicit

chrisbobbe committed:
BottomAppBar: Revert some doc changes

These might be good to make in a future PR, but they increase the
size of this PR's diff, and they're not on the path to fixing the
named issues.

chrisbobbe committed:
BottomAppBar: Shrink PR diff a bit

chrisbobbe committed:
BottomAppBar: Shrink PR diff some more

chrisbobbe commented on packages/flutter/test/material/bottom_app_bar_test.dart:
```diff
@@ -218,19 +284,18 @@ void main() {
                 ),
                 bottomNavigationBar: BottomAppBar(
                   color: Color(0xff0000ff),
+                  surfaceTintColor: Colors.transparent,
```
I add this `surfaceTintColor: Colors.transparent` in a few tests of the `BottomAppBar`'s base color, when the by-default surface tint color would make that more complicated.

<details>

<summary>The doc on `surfaceTintColor` seems wrong:</summary>

https://api.flutter.dev/flutter/material/BottomAppBar/surfaceTintColor.html
> If this is null, no overlay will be applied.

https://github.com/flutter/flutter/blob/5054b6e51/packages/flutter/lib/src/material/bottom_app_bar.dart#L193
```dart
final Color surfaceTintColor = widget.surfaceTintColor ?? babTheme.surfaceTintColor ?? defaults.surfaceTintColor!;
```

If it's null, then an overlay _will_ apply: one from the ambient `BottomAppBarTheme`, else from a hard-coded default for M2 or M3.

</details>

justinmc replied:
Could you file a new issue about the docs being wrong?

chrisbobbe replied:
Sure, I'll aim to do that today.

chrisbobbe replied:
#124535

chrisbobbe commented:
This is ready for another review; I've added commits following @HansMuller's comment at https://github.com/flutter/flutter/issues/123283#issuecomment-1481637735:

> [T]he notch should still work with `useMaterial3: true`.

The notch now does work in Material 3 (🎉), except that the default FAB shape doesn't have a `NotchedShape` written for it yet. That would be good to do.

| Before (`color` doubled; notch only in one layer) | After (`color` correct) |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/22248748/227042226-e7aeb8fc-c552-4a4e-80fb-5965253d18b4.png) | ![image](https://user-images.githubusercontent.com/22248748/227422568-7a695a42-6038-403a-b54a-0b0973de9712.png) |

gnprice commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -201,11 +203,7 @@ class _BottomAppBarState extends State<BottomAppBar> {
-      type: isMaterial3 ? MaterialType.canvas : MaterialType.transparency,
-      elevation: elevation,
-      color: isMaterial3 ? effectiveColor : null,
-      surfaceTintColor: surfaceTintColor,
-      shadowColor: shadowColor,
+      type: MaterialType.transparency,
       child: SafeArea(child: child),
```
I like how much simpler this gets!

gnprice reviewed (approved):
Cool, thanks for the revision. The new version LGTM.

gnprice commented:
Looks like you'll need to rebase:
![image](https://user-images.githubusercontent.com/28173/227424995-766edda8-851a-4a1c-94aa-c6bc1dbc4b16.png)

(I think a merge would probably also do it.)

chrisbobbe committed:
Merge branch 'main' into wip-bottom-app-bar-notch-material-3

chrisbobbe commented:
Thanks for the review!

> (I think a merge would probably also do it.)

Done.

chrisbobbe committed:
BottomAppBar tests: Fix an incomplete comment

gnprice commented:
Google testing failed. This is likely the same failures @justinmc mentioned above:

> By the way, the Google testing failures look like there's a small pixel difference at the right edge of the BottomAppBar in some image diff tests. Likely it's intended, but I'll take another look after the PR is updated.

Customer testing also failed, in a way unrelated to this PR: #123385

justinmc commented:
@chrisbobbe Can you merge master into this PR?  That will fix the customer tests and update the Google tests so I can take a look.

chrisbobbe committed:
Merge remote-tracking branch 'upstream/main'

chrisbobbe commented:
Sure! Done.

justinmc commented:
Thanks, looks like the customer tests are fixed.  I'll check on the Google tests next week.

justinmc commented:
The Google test failure looks unrelated... Can you try pushing another merge commit?  Sorry for all the headache with this.

chrisbobbe committed:
Merge remote-tracking branch 'upstream/main'

chrisbobbe commented:
Sure, no problem! Done.

chrisbobbe commented:
Hmm, Google testing is still showing as "pending" after 7+ hours. Could be #123289?

This time I'm going to try and see if I can update the branch without leaving GitHub's UI. There's a helpful-looking "Update branch" button that I'm going to click now. (If that does something wrong, I can always reset the branch back to b084982, my merge commit from earlier today.)

chrisbobbe committed:
Merge branch 'master' into pr-bottom-app-bar-notch-material-3

HansMuller commented:
@XilaiZhang - apparently this PR has failed "Google Testng" however the details link doesn't reveal anything: "There is no unmerged PR.". I tried the "Test a PR" button to start another test and that failed with:

![Screenshot 2023-04-04 at 8 47 48 AM](https://user-images.githubusercontent.com/1377460/229847021-2ba681a6-49c7-4ff6-8c06-6642adef4add.png)

How can we move the process along?

XilaiZhang commented:
temp workaround: the test failures are at test/521647510 and the cl is cl/521647510

Thanks for reporting the issue! will keep track of it in b/260764527

HansMuller commented:
The Google testing failures don't appear to have anything to do with this PR. Will try updating the branch...

HansMuller committed:
Merge branch 'master' into pr-bottom-app-bar-notch-material-3

justinmc reviewed (approved):
LGTM 👍 

I think that Google testing is broken right now, so we'll have to try pushing another merge commit in a day or so when it's fixed.

chrisbobbe commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -33,7 +33,8 @@ import 'theme.dart';
 ///
 /// {@tool dartpad}
-/// This example shows the [BottomAppBar], which can be configured to have a notch using the
+/// This example shows the [BottomAppBar], which (in Material 2)
```
Ah, this is already addressed in this revision (I suspect your review-draft crossed with my sending that new revision). I've updated the PR description to match this revision. 🙂

justinmc replied:
Thanks, yeah this review was messed up somehow...

chrisbobbe commented:
Thanks for the review!

> I think that Google testing is broken right now, so we'll have to try pushing another merge commit in a day or so when it's fixed.

Sure, I'll try that now.

chrisbobbe committed:
Merge branch 'master' into pr-bottom-app-bar-notch-material-3

justinmc commented:
Thanks for pushing the merge commit.  It looks like there are real Google test failures, but they are just nearly invisible image diffs due to the changes here.  I've accepted the changes and should be able to merge this soon.

gnprice commented:
Thanks @justinmc and everyone for your help in taking care of Google testing!

chrisbobbe commented:
Yeah, thanks all! 🎉
